include/boost/optional/detail/optional_trivially_copyable_base.hpp

```diff
@@ -39,7 +39,7 @@ class tc_optional_base : public optional_tag
 
     tc_optional_base ( none_t )
       :
-      m_initialized(false) {}
+      m_initialized(false), m_storage() {}
```

Comment:
This adds a write that will never be read.

Reply:
Ah, I understand now. I thought you were still using the `union` here as a form of aligned_storage.

There is a write here because `m_storage` is actually a `T`, not a `union { T t_ }`

You know what this means? It means that the `maybe-uninitialized` warning truly is genuine then!

If an optional<tc> is constructed with none and then copied, we're technically copying uninitialized values.

My first instinct would be to simply update this code to store a `union` but I'm not sure what the goals of this trivially copyable base specialization is.

Reply:
Alright, I played around with the code for like 30 minutes or so. Because we only use this for scalar types, I'm not sure that just doing a store here is the worst thing in the world. Under sufficient optimizations it probably gets eliminated.

Reply:
I think I have an idea:
```cpp
struct zst{};
union tc_storage { T t_; zst z_; };

  private :

    bool m_initialized ;
    tc_storage m_storage ;
```

Then whenever we need default constructibility, we just do: `m_storage.z_={};`

Exactly like this:
```cpp
tc_optional_base ( none_t )
  :
  m_initialized(false) {
    m_storage.z_={};
  }
```